### DISCUSSION TITLE
How to spy on declared functions with `sinon` without passing extra arguments

### DISCUSSION THREAD

**Discussion by GJZwiers** [denoland/deno#12603] [Q&A]
I have a function A that calls another function B:
```typescript
function A() {
  B();
}

function B() {

}
```

And I want to write a test that wraps function `B` using `sinon` and makes some assertions, such as if `B` has been called after invoking `A`:

```typescript
const spy = sinon.spy(B);

// do test, call A somewhere

assertEquals(spy.called, true);
```
The only way this works is if I add a parameter to `A` so I can replace `B` with the spy function for the test:

```typescript
function A(fn) {
  fn();
}

A(B); // and A(spy) in the test
```

but in reality `A` is more complex and I don't really want to add parameters purely for testing. In other JavaScript environments `B` can be accessed from a global such as `window`, like in the example from [this post](https://stackoverflow.com/a/29801191/12769288):

```javascript
function bar(x,y) {
  console.debug(x, y);
}
function foo(z) {
  bar(z, z+1);
}
// Spy on the function "bar" of the global object.
var spy = sinon.spy(window, "bar");
```

Is there an equivalent of this in Deno? I've tried `window` and `self`, but those error saying `B` is undefined on those objects. Also in my project `B` is first imported from another file, and I've tried `import * as foo from "my_file.ts"` and then `sinon.spy(foo, 'B')`, but in that case `sinon` errors saying `ES modules cannot be spied`. Can I spy on `B` without having to rewrite `A` to take an extra parameter?

**Comment by GJZwiers** [comment]
Here is what I ended up doing, in case anyone finds it useful. Add the functions `A` and `B` to an object and have `A` call `B` from the object:

```ts
// foo.ts

function A() {
  funcs.B();
}

function B() {

}

const funcs = {
  A,
  B,
};
```

Export `funcs` to a `.test` file and wrap it with `sinon` to spy on `B`:

```ts
// foo.test.ts

import sinon from "https://cdn.skypack.dev/sinon@12.0.1?dts";
import { funcs } from "./foo.ts";

const spy = sinon.spy(funcs, "B");

// Deno.test something that calls A, then assert that B was called once:
assertEquals(spy.getCalls().length, 1);
```

**Comment by bartlomieju** [comment]
@GJZwiers I suggest to upstream this solution to https://github.com/denoland/manual

**Reply by GJZwiers** [reply]
Sounds good! I will see if I can find some time for it in the coming weeks :)
